// @shopify/app/dist/cli/commands/app/config/link.js
var _a;
import { appFlags } from '../../../flags.js';
import link from '../../../services/app/config/link.js';
import Command from '../../../utilities/app-command.js';
import { Flags } from '@oclif/core';
import { globalFlags } from '@shopify/cli-kit/node/cli';
class ConfigLink extends Command {
    async run() {
        const { flags } = await this.parse(_a);
        const options = {
            directory: flags.path,
            apiKey: flags['client-id'],
        };
        await link(options);
    }
}
_a = ConfigLink;
ConfigLink.summary = 'Fetch your app configuration from the Partner Dashboard.';
ConfigLink.descriptionWithMarkdown = `Pulls app configuration from the Partner Dashboard and creates or overwrites a configuration file. You can create a new app with this command to start with a default configuration file.

  For more information on the format of the created TOML configuration file, refer to the [App configuration](https://shopify.dev/docs/apps/tools/cli/configuration) page.
  `;
ConfigLink.description = _a.descriptionWithoutMarkdown();
ConfigLink.flags = {
    ...globalFlags,
    ...appFlags,
    'client-id': Flags.string({
        hidden: false,
        description: 'The Client ID of your app.',
        env: 'SHOPIFY_FLAG_CLIENT_ID',
    }),
};
export default ConfigLink;
