// @shopify/app/dist/cli/commands/app/function/run.js
var _a;
import { functionFlags, inFunctionContext } from '../../../services/function/common.js';
import { runFunctionRunner } from '../../../services/function/build.js';
import { appFlags } from '../../../flags.js';
import Command from '@shopify/cli-kit/node/base-command';
import { globalFlags } from '@shopify/cli-kit/node/cli';
import { Flags } from '@oclif/core';
class FunctionRun extends Command {
    async run() {
        const { flags } = await this.parse(_a);
        await inFunctionContext({
            path: flags.path,
            configName: flags.config,
            callback: async (_app, ourFunction) => {
                await runFunctionRunner(ourFunction, {
                    json: flags.json,
                    input: flags.input,
                    export: flags.export,
                });
            },
        });
    }
}
_a = FunctionRun;
FunctionRun.summary = 'Run a function locally for testing.';
FunctionRun.descriptionWithMarkdown = `Runs the function from your current directory for [testing purposes](https://shopify.dev/docs/apps/functions/testing-and-debugging). To learn how you can monitor and debug functions when errors occur, refer to [Shopify Functions error handling](https://shopify.dev/docs/api/functions/errors).`;
FunctionRun.description = _a.descriptionWithoutMarkdown();
FunctionRun.flags = {
    ...globalFlags,
    ...appFlags,
    ...functionFlags,
    input: Flags.string({
        char: 'i',
        description: 'The input JSON to pass to the function. If omitted, standard input is used.',
        env: 'SHOPIFY_FLAG_INPUT',
    }),
    export: Flags.string({
        char: 'e',
        hidden: false,
        description: 'Name of the wasm export to invoke.',
        default: '_start',
        env: 'SHOPIFY_FLAG_EXPORT',
    }),
    json: Flags.boolean({
        char: 'j',
        hidden: false,
        description: 'Log the run result as a JSON object.',
        env: 'SHOPIFY_FLAG_JSON',
    }),
};
export default FunctionRun;
