// @shopify/app/dist/cli/commands/app/function/typegen.js
var _a;
import { functionFlags, inFunctionContext } from '../../../services/function/common.js';
import { buildGraphqlTypes } from '../../../services/function/build.js';
import { appFlags } from '../../../flags.js';
import Command from '@shopify/cli-kit/node/base-command';
import { globalFlags } from '@shopify/cli-kit/node/cli';
import { renderSuccess } from '@shopify/cli-kit/node/ui';
class FunctionTypegen extends Command {
    async run() {
        const { flags } = await this.parse(_a);
        await inFunctionContext({
            path: flags.path,
            configName: flags.configName,
            callback: async (app, ourFunction) => {
                await buildGraphqlTypes(ourFunction, { stdout: process.stdout, stderr: process.stderr, app });
                renderSuccess({ headline: 'GraphQL types generated successfully.' });
            },
        });
    }
}
_a = FunctionTypegen;
FunctionTypegen.summary = 'Generate GraphQL types for a JavaScript function.';
FunctionTypegen.descriptionWithMarkdown = `Creates GraphQL types based on your [input query](https://shopify.dev/docs/apps/functions/input-output#input) for a function written in JavaScript.`;
FunctionTypegen.description = _a.descriptionWithoutMarkdown();
FunctionTypegen.flags = {
    ...globalFlags,
    ...appFlags,
    ...functionFlags,
};
export default FunctionTypegen;
