// @shopify/app/dist/cli/services/release/version-diff.js
import { renderError } from '@shopify/cli-kit/node/ui';
import { AbortSilentError } from '@shopify/cli-kit/node/error';
export async function versionDiffByVersion(apiKey, version, developerPlatformClient) {
    const versionDetails = await versionDetailsByTag(apiKey, version, developerPlatformClient);
    const input = {
        apiKey,
        versionId: versionDetails.id,
    };
    const { app: { versionsDiff }, } = await developerPlatformClient.appVersionsDiff(input);
    return { versionsDiff, versionDetails };
}
async function versionDetailsByTag(apiKey, version, developerPlatformClient) {
    try {
        const input = {
            apiKey,
            versionTag: version,
        };
        const { app: { appVersion }, } = await developerPlatformClient.appVersionByTag(input);
        return appVersion;
    }
    catch (err) {
        renderError({
            headline: "Version couldn't be released.",
            body: ['Version', { userInput: version }, 'could not be found.'],
        });
        throw new AbortSilentError();
    }
}
