// @shopify/app/dist/cli/services/release.js
import { ensureReleaseContext } from './context.js';
import { configExtensionsIdentifiersBreakdown, extensionsIdentifiersReleaseBreakdown, } from './context/breakdown-extensions.js';
import { deployOrReleaseConfirmationPrompt } from '../prompts/deploy-release.js';
import { renderError, renderSuccess, renderTasks } from '@shopify/cli-kit/node/ui';
import { AbortSilentError } from '@shopify/cli-kit/node/error';
export async function release(options) {
    const { developerPlatformClient, app, remoteApp } = await ensureReleaseContext(options);
    const { extensionIdentifiersBreakdown, versionDetails } = await extensionsIdentifiersReleaseBreakdown(developerPlatformClient, remoteApp.apiKey, options.version);
    const configExtensionIdentifiersBreakdown = await configExtensionsIdentifiersBreakdown({
        developerPlatformClient,
        apiKey: remoteApp.apiKey,
        localApp: app,
        remoteApp,
        versionAppModules: versionDetails.appModuleVersions.map((appModuleVersion) => ({
            ...appModuleVersion,
            ...(appModuleVersion.config ? { config: JSON.parse(appModuleVersion.config) } : {}),
        })),
        release: true,
    });
    const confirmed = await deployOrReleaseConfirmationPrompt({
        configExtensionIdentifiersBreakdown,
        extensionIdentifiersBreakdown,
        appTitle: remoteApp.title,
        release: true,
        force: options.force,
    });
    if (!confirmed)
        throw new AbortSilentError();
    const variables = {
        apiKey: remoteApp.apiKey,
        appVersionId: versionDetails.id,
    };
    const tasks = [
        {
            title: 'Releasing version',
            task: async (context) => {
                context.appRelease = await developerPlatformClient.release(variables);
            },
        },
    ];
    const { appRelease: { appRelease: release }, } = await renderTasks(tasks);
    const linkAndMessage = [
        { link: { label: versionDetails.versionTag, url: versionDetails.location } },
        versionDetails.message ? `\n${versionDetails.message}` : '',
    ];
    if (release.userErrors?.length > 0) {
        renderError({
            headline: "Version couldn't be released.",
            body: [
                ...linkAndMessage,
                `${linkAndMessage.length > 0 ? '\n\n' : ''}${release.userErrors.map((error) => error.message).join(', ')}`,
            ],
        });
    }
    else {
        renderSuccess({
            headline: 'Version released to users.',
            body: linkAndMessage,
        });
    }
}
